File: exercises/feature-effects/rsrc/code/ale.py
import sys
sys.path.insert(0, ".")

import pandas as pd
import numpy as np

from utils.dataset import Dataset
from utils.styled_plot import plt


def get_bounds(X, s, n_intervals=100):
    """
    Calculates interval bounds given a X's min and max values.
    Only values in the s-th column are considered.
    
    Parameters:
        X (np.array with shape (num_instances, num_features)): Input data.
        s (int): Index of the feature x_s.
        n_intervals (int): Number of intervals.
        
    Returns:
        bounds (list): Values of bounds with `n_intervals`+1 entries.
    """
    
    return None


def calculate_ale(model, X, s, n_intervals=100, centered=False):
    """
    Compute the accumulated local effect of a numeric continuous feature.
    
    Parameters:
        model: Classifier which has a predict method.
        X (np.array with shape (num_instances, num_features)): Input data.
        s (int): Index of the feature x_s.
        n_intervals (int): How many intervals should be used.
        centered (bool): Whether to return uncentered or centered ALE.
        
    Returns:
        bounds (list): See function `get_bounds`.
        ale (list with `n_intervals` entries): Values of ALE.
    """

    return None, None


def prepare_ale(model, X, s, n_intervals=100, centered=False):
    """
    Uses `calculate_ale` to prepare x and y data, which can be used
    by matplotlib directly.
    
    Parameters:
        model: Classifier which has a predict method.
        X (np.array with shape (num_instances, num_features)): Input data.
        s (int): Index of the feature x_s.
        n_intervals (int): How many intervals should be used.
        centered (bool): Whether to return uncentered or centered ALE.  
        
    Returns:
        x (list or 1D np.ndarray): Centers of two bounds. `n_intervals` entries.
        y (list or 1D np.ndarray): ALE values. `n_intervals` entries.
    """

    return None, None
        
        
if __name__ == "__main__":
    dataset = Dataset("wheat_seeds", [5,6,7], [2], normalize=True, categorical=False)
    (X_train, y_train), (X_test, y_test) = dataset.get_data()
    
    from sklearn import ensemble
    model = ensemble.RandomForestRegressor(random_state=0)
    model.fit(X_train, y_train)
    X = dataset.X
    s = 1
    
    print("Run `get_bounds` ...")
    print(get_bounds(X, 0, n_intervals=4))
    
    num_samples = X.shape[0]
    n_intervals = [int(num_samples/16), int(num_samples/8), int(num_samples/4)]
    num_features = len(dataset.input_ids)
    
    print("Plot ...")
    plt.subplots(cols=len(n_intervals), rows=2, sharey=True)
    for row_id, centered in enumerate([False, True]):
        for col_id, k in enumerate(n_intervals):
                
            plt.subplot(2, len(n_intervals), col_id+1+len(n_intervals)*row_id)
            plt.title(f"{k} bins")
            plt.xlabel(dataset.get_input_labels(s))
            if col_id == 0:
                centered_text = ""
                if not centered:
                    centered_text = "Uncentered"
                        
                plt.ylabel(f"{centered_text} ALE [{dataset.get_output_label()}]")
                    
            x, y = prepare_ale(model, X, s, n_intervals=k, centered=centered)
            plt.plot(x, y, linewidth=0.5)
            
    plt.tight_layout(pad=0.75)
    plt.show()

File: exercises/03_feature_effects/code/ale.R
library("docstring")

get_bounds <- function(X, s, n_intervals = 100) {
  
  #' Calculates interval bounds given a X's min and max values.
  #' Only values in the s-th column are considered.
  #
  #' @param X (data.frame): Input data.
  #' @param s (integer(1)): Index of the feature x_s.
  #' @param n_intervals (integer(1)): Number of intervals.
  #'
  #' @return bounds (vector): Values of bounds with `n_intervals`+1 entries.
  #'

  return(NULL)
}

calculate_ale <- function(model, X, s, n_intervals = 100, centered = FALSE) {
  
  #' Compute the accumulated local effect of a numeric continuous feature.
  #'
  #' @param model: Classifier which has a predict method.
  #' @param X (data.frame): Input data.
  #' @param s (integer(1)): Index of the feature x_s.
  #' @param n_intervals (integer(1)): How many intervals should be used.
  #' @param centered (logical(1)): Whether to return uncentered or centered ALE.
  #'
  #' @return a list with the following elements: 
  #' bounds (vector): See function `get_bounds`.
  #' ale (vector): Values of ALE.

  return(list(bounds = NULL, ale = NULL))
}

prepare_ale <- function(model, X, s, n_intervals = 100, centered = TRUE) {
  #' Uses `calculate_ale` to prepare x and y data, which can be used
  #' by ggplot2 directly.
  #'
  #' @param model: Classifier which has a predict method.
  #' @param X (data.frame): Input data.
  #' @param s (integer(1)): Index of the feature x_s.
  #' @param n_intervals (integer(1)): Number of intervals.
  #' @param centered (logical(1)): Whether to return uncentered or centered ALE.
  #'
  #' @return data.frame with two columns and `n_intervals` rows 
  #'     * x: Centers of two bounds.
  #'     * y: ALE values.
  #'
  
  return(NULL)
  
}

### A function to create the ALE plot from the previous generated objects.
ale_plot <- function() {
  library(randomForest)
  library(ggplot2)
  
  ### Set up your working directory using swd() and get the dataset file. 
  df <- read.csv(file = 'datasets/wheat_seeds.csv')
  
  ### Split the dataset to 70% train data and 30% test data.
  set.seed(100)
  train <- sample(nrow(df), 0.7 * nrow(df), replace = FALSE)
  trainData <- df[train, ]
  testData <- df[-train, ]
  
  ### Normalize the target to be between 0 and 1.
  min_max_norm <- function(x) {
    (x - min(x)) / (max(x) - min(x))
  }
  
  trainData$Type <- min_max_norm(trainData$Type)
  
  ### Build a Random Forest model.
  model <- randomForest(Type ~ ., data = trainData, mtry = 4, importance = TRUE)
  
  ### Use the first feature from the dataset and set up
  ### 4 intervals to test the get_bounds function
  bounds <- get_bounds(df, 1, 4)
  
  ### Test the calculate_ale function, with centered = FALSE
  uncentered_ale <- calculate_ale(model, df, 1, 4, FALSE)
  
  ### Test the prepare_ale function, with centered = TRUE
  prepared_ale <- prepare_ale(model, df, 1, 4, TRUE)
  
  ggplot(data = prepared_ale, mapping = aes(x = x, y = y)) + 
    geom_line() + 
    geom_point()
}

### The warning message can be ignored here.
ale_plot()
